quickbrick tests and bug fixes

sbailey opened the pull request:
This PR
 * adds tests for quickbrick
 * fixes #161 (updating how exposure time is set with specsim v0.5)
 * fixes #162 (quickbrick crash)
 * closes #96 (tests for quickgen and quickbrick)

It also fixes some bugs that weren't previously known and were caught by the new unit tests:
 * --seed option wasn't working
 * truth table output was corrupted using py3

However:
  * The tests won't work on Travis until desihub/specsim#43 is addressed to reduce the specsim memory footprint or another workaround is found.  I've flagged these with `@unittest.skipIf('TRAVIS_JOB_ID' in os.environ)` wrappers.  Note that this also impacts the quickgen tests -- you can run them on a laptop with sufficient memory, but not on Travis.  In the meantime, we also won't get accurate coverage reports on coveralls.io for these.
  * The tests are slow (2.5 minutes for just the quickbrick tests).  Although quicksim is fast for N>>1 spectra in a single shot, generating the templates on the fly is not, and quicksim has a fairly large overhead that we encounter multiple times when comparing things like "did making the moon brighter make the spectra noiser?"
  * I'm not sure what the original intent of OBJTYPE vs. TRUE_OBJTYPE was for the truth table.  What I have in there currently at least runs, but it may need further modification.

Related: apologies for the noise I generated on issue #161 where I kept typing "specsim" when I meant "desisim".  @dkirkby and I meant the same thing but I kept typing the wrong thing and misreading the right thing that he was typing...

sbailey commented:
Any comments / complaints / suggestions?  Or "hold off I'll review this next week..."?

moustakas commented:
I hope to take a look this weekend!

sbailey commented:
Last call for comments.  This is becoming a blocking factor for other PR, e.g. #169 which is failing tests that are fixed in this PR.  I'll merge this afternoon with or without comments and we can retroactively fix stuff if that was a huge mistake.

dkirkby commented on py/desisim/test/test_quickbrick.py:
```diff
@@ -0,0 +1,127 @@
+    @unittest.skipIf('TRAVIS_JOB_ID' in os.environ, 'Skipping memory hungry quickbrick/specsim test on Travis')
+    def test_quickbrick_moon(self):
+        nspec = 2
+        cmd = "quickbrick --seed 1 -b brightmoon --objtype BRIGHT_MIX -n {} --outdir {} --moon-phase 0.1".format(nspec, self.testdir)
```
I would explicitly put the moon above the horizon here, instead of relying on a default.

dkirkby commented on py/desisim/test/test_quickbrick.py:
```diff
@@ -0,0 +1,127 @@
+    
+    #- Pick random test directory
+    def setUp(self):
+        self.testdir = 'test-{uuid}'.format(uuid=uuid1())
```
It might be cleaner and more portable to use the [tempfile module](https://docs.python.org/2/library/tempfile.html) here.

dkirkby commented on py/desisim/test/test_quickbrick.py:
```diff
@@ -0,0 +1,127 @@
+        brickfile = '{}/brick-b-{}.fits'.format(self.testdir, brickname)
+        truth = fits.getdata(brickfile, '_TRUTH')        
+        z = truth['TRUEZ']
+        self.assertTrue(np.all((0.1 <= z) & (z <= 0.2)))
```
Could also check that the median ivar is within some tolerance of the expected value, to catch any changes that significantly change the results.

dkirkby commented on py/desisim/scripts/quickbrick.py:
```diff
@@ -228,8 +233,9 @@ def main(args):
     meta.add_column(Column(targetids, name='TARGETID'))
 
+    # @sbailey asks: What was this supposed to do for OBJTYPE vs. TRUE_OBJTYPE?
     # Now add fixed-up OBJTYPE to the template meta table
```
Hopefully @moustakas can answer this.

moustakas replied:
Previously ```OBJTYPE``` was not included in the metadata table returned by ```desisim.templates```, but it was added as part of #132, which broke this part of the code.  We need this renaming of ```OBJTYPE--->TRUE_OBJTYPE``` for contaminants, however, for example, to identify objects with ```OBJTYPE='QSO_BAD'``` which are actually ```TRUE_OBJTYPE='STAR'```.  (Note that I didn't write this particular part of ```quickbrick``` but I believe that it's still needed downstream in the redshift fitting and classification portion of the pipeline.)

Long story short I think the line that's commented out can be removed safely.

moustakas commented:
Modulo @dkirkby's comments, this looks fine to merge.

sbailey commented:
Thanks for comments.
* I added options to ensure that moon is above horizon when testing `-b brightmoon` instead of relying upon defaults
* I added sanity check on median(error) vs. median(flux) for simulated spectra.  This test is intended to catch if something significantly changes; it isn't a detailed validation of the functional form of that.
* tempfile vs. uuid: we use this uuid method throughout multiple packages and tests.  I'm leaving as-is unless we find a specific problem with it that `tempfile` would fix.

I believe this addresses all comments and I will merge after tests pass on Travis.
